### PR TITLE
Shiny (Python) ML training example

### PR DESCRIPTION
**Pull request by neelasha23**
Add Shiny example

----
📚 Documentation preview 📚: https://ploomber-doc--105.org.readthedocs.build/en/105/

### CODE REVIEW

File: doc/_toc.yml [hunk 1]
```diff
@@ -29,6 +29,7 @@ parts:
       - file: apps/panel
       - file: apps/chainlit
       - file: apps/gradio
+      - file: apps/shiny-core
```
Comment (edublancas) [inline doc/_toc.yml:32]: let's call this shiny, I'm unfamiliar with the "shiny core" term and other might be confused as well

File: doc/apps/shiny-core.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# Shiny Core
```
Comment (edublancas) [inline doc/apps/shiny-core.md:1]: rename al references from "shiny core" to "shiny"

File: examples/docker/shiny-core/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
+# Model monitoring application
+
+This app is built on Shiny and allows you to visualize model training accuracy scores in realtime in 2 steps:
+
+* Train multiple models with different parameters locally, and store the accuracy scores to a remote database at the end of every epoch.
+* The visualisation application dynamically detects changes in the remote database and generates a plot as the training progresses.
+
+## Connecting to MongoDB Cloud
+
+For the purpose of this tutorial we will use the `MongoDB Atlas` cloud database. Let's see the steps for creating a cluster:
```
Comment (edublancas) [inline examples/docker/shiny-core/README.md:10]: mention that this service allows you to host a database for free

File: examples/docker/shiny-core/README.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+![](./static/ip-access.png)
+
+Once you have completed the access rules setup, select the `CONNECT` button on the `Overview` page to get the connection details. Specifically, copy the connection string and replace the `URI` variable value in `app-core.py`, `train_one.py` and `train_two.py`
+
+![](./static/connect.png)
+
+## Deploy application
+
+Create a zip file using `Dockerfile`, `app-core.py` and `requirements.txt` and deploy this file to Ploomber cloud.
```
Comment (edublancas) [inline examples/docker/shiny-core/README.md:34]: add a link to the documentation that shows how to deploy shiny apps

File: examples/docker/shiny-core/README.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+## Train models
+
+We will use the [MNIST dataset](https://www.tensorflow.org/datasets/catalog/mnist) and train two different models on the data. The models differ in the number of units in the `Dense` layer.
+The training scripts write the accuracy values at the end of every epoch to the MongoDB database in a collection called `accuracy_scores`.
+
+First create a virtual environment and install the required packages:
+
+```bash
+pip install tensorflow scikit-learn "pymongo[srv]"
```
Comment (edublancas) [inline examples/docker/shiny-core/README.md:45]: so we need scikit-learn? I don't see it imported

File: examples/docker/shiny-core/README.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+```
+
+Then run the training scripts parallely.:
```
Comment (edublancas) [inline examples/docker/shiny-core/README.md:48]:
```suggestion
Then run the training scripts parallely:
```

File: examples/docker/shiny-core/README.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+```bash
+python train_one.py & python train_two.py & wait
```
Comment (edublancas) [inline examples/docker/shiny-core/README.md:51]: `&` is not available on windows, you can tell people to open two terminal windows and execute the commands on each one.

also, the scripts are pretty much the same, what changes between those two? It'd be better to have one script and just pass an argument:

```
python train.py --some-arg value
python train.py --some-arg another
```

File: examples/docker/shiny-core/app-core.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
+from __future__ import annotations
```
Comment (edublancas) [inline examples/docker/shiny-core/app-core.py:1]: this is only needed for Python 2, I don't think it's used in this script

File: examples/docker/shiny-core/app-core.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+from pymongo.mongo_client import MongoClient
+from pymongo.server_api import ServerApi
+from shiny import App, Inputs, Outputs, Session, reactive, render, ui
+
+THRESHOLD_MID = 0.9
+THRESHOLD_MID_COLOR = "rgb(0, 137, 26)"
+THRESHOLD_LOW = 0.5
+THRESHOLD_LOW_COLOR = "rgb(193, 0, 0)"
+
+URI = "CONNECTION_STRING"
```
Comment (edublancas) [inline examples/docker/shiny-core/app-core.py:17]: read this from and env variable

File: examples/docker/shiny-core/README.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+Select the `Cloud Environment` and open the `Network Access Page` in a new tab.
+
+![](./static/network-access.png)
+
+Now select `ALLOW ACCESS FROM ANYWHERE` and confirm to allow all IP addresses to connect to the cluster.
+
+![](./static/ip-access.png)
+
+Once you have completed the access rules setup, select the `CONNECT` button on the `Overview` page to get the connection details. Specifically, copy the connection string and replace the `URI` variable value in `app-core.py`, `train_one.py` and `train_two.py`
```
Comment (edublancas) [inline examples/docker/shiny-core/README.md:28]: show how to set and environment variable and modify the scripts to read from there

File: examples/docker/shiny-core/app-core.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
+from __future__ import annotations
+
+import pandas as pd
```
Comment (edublancas) [inline examples/docker/shiny-core/app-core.py:3]: rename this to app.py

**Review by edublancas — CHANGES_REQUESTED** [review]
please deploy the app so I can test it

File: examples/docker/shiny-core/README.md (hunk 5 above)
Comment (neelasha23) [reply examples/docker/shiny-core/README.md:45]: It's there in train_one.py and train_two.py

File: doc/_toc.yml (hunk 1 above)
Comment (neelasha23) [reply doc/_toc.yml]: [fixed](https://github.com/ploomber/doc/blob/6318fe9c144499deef0d0550c0626f73f6942aaf/doc/_toc.yml#L32)

File: doc/apps/shiny-core.md (hunk 2 above)
Comment (neelasha23) [reply doc/apps/shiny-core.md]: [fixed](https://github.com/ploomber/doc/blob/6318fe9c144499deef0d0550c0626f73f6942aaf/doc/apps/shiny.md?plain=1#L1)
Changed all references

File: examples/docker/shiny-core/README.md (hunk 3 above)
Comment (neelasha23) [reply examples/docker/shiny-core/README.md]: [fixed](https://github.com/ploomber/doc/pull/105/files#diff-883b5ca3546bed48155f3709e49ebe5d5b987079fdc8b9c9d9f930a943512616R10-R11)

File: examples/docker/shiny-core/README.md (hunk 4 above)
Comment (neelasha23) [reply examples/docker/shiny-core/README.md]: [fixed](https://github.com/ploomber/doc/pull/105/files#diff-883b5ca3546bed48155f3709e49ebe5d5b987079fdc8b9c9d9f930a943512616R40-R41)

File: examples/docker/shiny-core/README.md (hunk 6 above)
Comment (neelasha23) [reply examples/docker/shiny-core/README.md]: Sentenced replaced by : https://github.com/ploomber/doc/pull/105/files#diff-883b5ca3546bed48155f3709e49ebe5d5b987079fdc8b9c9d9f930a943512616R59-R60

File: examples/docker/shiny-core/README.md (hunk 7 above)
Comment (neelasha23) [reply examples/docker/shiny-core/README.md]: [fixed](https://github.com/ploomber/doc/pull/105/files#diff-883b5ca3546bed48155f3709e49ebe5d5b987079fdc8b9c9d9f930a943512616R59-R67)

File: examples/docker/shiny-core/app-core.py (hunk 8 above)
Comment (neelasha23) [reply examples/docker/shiny-core/app-core.py]: [Removed](https://github.com/ploomber/doc/pull/105/files#diff-c99010220192b232032cc417d9e26e799aa65a2dd2fbb5dd79b5595548293670R1-R3) from `app.py`

File: examples/docker/shiny-core/app-core.py (hunk 9 above)
Comment (neelasha23) [reply examples/docker/shiny-core/app-core.py]: [fixed](https://github.com/ploomber/doc/pull/105/files#diff-c99010220192b232032cc417d9e26e799aa65a2dd2fbb5dd79b5595548293670R18-R19)

File: examples/docker/shiny-core/README.md (hunk 10 above)
Comment (neelasha23) [reply examples/docker/shiny-core/README.md]: [fixed](https://github.com/ploomber/doc/pull/105/files#diff-883b5ca3546bed48155f3709e49ebe5d5b987079fdc8b9c9d9f930a943512616R42-R44)

[fixed](https://github.com/ploomber/doc/pull/105/files#diff-883b5ca3546bed48155f3709e49ebe5d5b987079fdc8b9c9d9f930a943512616R29)

File: examples/docker/shiny-core/app-core.py (hunk 11 above)
Comment (neelasha23) [reply examples/docker/shiny-core/app-core.py]: Renamed

**Comment by neelasha23** [comment]
Addressed the comments. App is deployed here: https://crimson-glade-4646.ploomberapp.io/
After setting the `MONGODB_CONNECTION_URI` env variable in local u can run the two scripts as mentioned in Readme: `python train.py --model model_1 --units 128` and `python train.py --model model_2 --units 12`
@edublancas

**Review by lfunderburk — COMMENTED** [review]
Hi @neelasha23 

Took a look at the app - but the plot is not showing up
Screenshot 2024-02-02 at 4 00 51 PM

**Comment by neelasha23** [comment]
> Hi @neelasha23
> 
> Took a look at the app - but the plot is not showing up <img alt="Screenshot 2024-02-02 at 4 00 51 PM" width="2000" 

I have cleared the db because it's a realtime monitoring. you need to connect to ur mongodb account and run training scripts from local and view the plot in the dashboard. have added instructions in the README @lfunderburk

File: examples/docker/shiny/train.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+my_collection = db["accuracy_scores"]
+
+
+class TrackLossandAccuracyCallback(tf.keras.callbacks.Callback):
+
+    def __init__(self, model_name):
+        self.model_name = model_name
+
+    def on_epoch_end(self, epoch, logs=None):
+        my_collection.insert_one({'model': self.model_name, 'score': logs["accuracy"], 'timestamp': datetime.now()})
```
Comment (edublancas) [inline examples/docker/shiny/train.py:33]: add a print statement here so users can compare what they see in the terminal with what they see in the app

**Review by edublancas — CHANGES_REQUESTED** [review]

**Comment by edublancas** [comment]
the polling logic is off: I tried the app many times (both the ploomberapp.io one and locally) and I only got it one time to work. I debugged it by printing the value of `last_modified` and realized that even if I started training, the returned value was the same (I tried refreshing the browser after I started training).

I changed the `last_modified` function to this:

```python
def last_modified():
    n = my_collection.count_documents({})
    return n
```

and it worked better: as soon as I started training, the plot started updating. but I found another issue: the values in the plot did not match the values printed from the terminal and when starting to train another model, the plot didn't show a new line:


https://github.com/ploomber/doc/assets/989250/f9ba9706-86bb-4239-b8b2-794dcc14b485

**Comment by neelasha23** [comment]
I haven't been able to reproduce the timestamp not getting updated issue. Did you try with my account? I had selected an Indian server, so I'm doubting if that's causing a lag.
Can you please try once with a new account? If it still doesn't work I'll figure out what's happening
@edublancas 


https://github.com/ploomber/doc/assets/29096135/b1053d2b-a84b-497f-82d1-e51a86f88144

**Comment by edublancas** [comment]
@neelasha23 the problem is most likely the `datetime.now()`, such function returns the timestamp in your local timezone.

but there's no need to reproduce it since I already provided a fix:

```python
def number_of_observations():
    n = my_collection.count_documents({})
    return n
```

this function will trigger an update as soon as there is a new element in the collection

File: examples/docker/shiny/train.py (hunk 12 above)
Comment (neelasha23) [reply examples/docker/shiny/train.py]: [fixed](https://github.com/ploomber/doc/pull/105/files#diff-de90239477773a3e77cf5636a10f5db4b7b1bab1331462d3f5330d480173cf4cR35-R36)

**Comment by neelasha23** [comment]
Fixed the polling function: https://github.com/ploomber/doc/pull/105/files#diff-c99010220192b232032cc417d9e26e799aa65a2dd2fbb5dd79b5595548293670R33-R38

Although I did not completely understand why polling the timestamp column would be an issue. Even if datetime generates in local timezone it will still get accurately stored in MongoDB. And the polling function keeps checking this stored timestamp to detect if any new row has been added. You also mentioned that even after changing the polling function the timestamps in the plot are not matching the training ones , but I don't see that happening in my local. Which is why I feel maybe there's a lag between the training and the storage/retrieval process in your case? 
@edublancas 


https://github.com/ploomber/doc/assets/29096135/49d4b758-bd3e-4927-9046-1ba8e7b72fe5
